RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,750 @@
+# F# RFC FS-1063 - Support let! .. and... for applicative functors
+
+The design suggestion [Support let! .. and... for applicative functors](https://github.com/fsharp/fslang-suggestions/issues/579) has been marked "approved in principle".
+This RFC covers the detailed proposal for this suggestion.
+
+* [x] [Approved in principle](https://github.com/fsharp/fslang-suggestions/issues/579#event-1345361104) & [prioritised](https://github.com/fsharp/fslang-suggestions/issues/579#event-1501977428)
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/579)
+* [x] [Discussion](https://github.com/fsharp/fslang-design/issues/335)
+* [ ] Implementation: [In progress](https://github.com/Microsoft/visualfsharp/pull/5696)
+
+# Summary
+[summary]: #summary
+
+Extend computation expressions to support applicative functors via a new `let! ... and! ... return ...` syntax.
+
+# Motivation
+[motivation]: #motivation
+
+Applicative functors (or just "applicatives", for short) have been growing in popularity as a way to build applications and model certain domains over the last decade or so, since McBride and Paterson published [Applicative Programming with Effects](http://www.staff.city.ac.uk/~ross/papers/Applicative.html). Applicatives are now reaching a level of popularity within the community that supporting them with a convenient and readable syntax, as we do for monads, makes sense.
+
+With applicative computation expressions, we can write more computations than before (there are more contexts which meet the requirements for applicative computation expressions than the existing monadic ones), and we can write more efficient computations (the requirements of applicatives rule out needing to support some potentially expensive operations).
+
+For example, [Pauan points out](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-310799948) that we can write a computation expression for `Observable`s that [avoids unnecessary resubscriptions](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-310854419):
+
+```fsharp
+// Outputs a + b, which is recomputed every time foo or bar outputs a new value,
+// avoiding any unnecessary resubscriptions
+observable {
+  let! a = foo
+  and! b = bar
+  return a + b
+}
```

Comment:
For computation expressions, this is a very interesting case. What is meant by `let! ... and! ...`? This looks, on first glance, like an example of `Observable.zip`. However, what's to stop someone from implementing this as a pre-defined `Observable.Join`? I think both are valid implementations for a computation expression, depending on use case.

Reply:
This isn't something that I know much about or have given a vast amount of thought to, so I could be wrong on this: Nothing special here apart from there being a convenient syntax for doing something like `Observable.zip` but for any number of `Observable`s. Note also how the new syntax forces names to be given to each parameter right next to where the argument is supplied, making it harder to accidentally swap `Observable`s being passed in which are of the same type.

I don't know what you would necessarily mean by `Observable.Join`.

I think the point here is that I perhaps wrote a slightly misleading description in the RFC. It's not that something sensible is inexpressible now (e.g. via `Observable.zip`), but more that if you wanted to use some nice computation expression syntax for it, you'd need to pay an annoying performance cost.

Reply:
I'll look to make that clearer in the RFC.

Reply:
I've subtly changed the wording around this in an attempt to clarify that it wasn't impossible to implement applicatives before, but that this is just convenient syntax for them.